spec/features/admin/customers_spec.rb

```diff
@@ -148,6 +136,24 @@
             expect(page).to have_content "$0.00"
           end
         end
+
+        # this adds the additional payments to an order, and checks for customer balance
+        it "updates customer balance with additional negative payments, on an order" do
+          payment2
```

Comment:
again, a `let!` would make this unnecessary.

Reply:
I think this variable needs to be lazily evaluated: if evaluated in the beginning of the test case, then the payment is not added to the order.

To put it in another way, calling it in the moment we evaluate the test case allows to actually _add_ a payment, and compare the balance of the customer _before_ this payment was added, to the value _after_ this variable was called. I would add this comment above the calling of the variable:

`# payment2 needs to be lazily defined to be called here, as way to add a payment to order1`

Reply:
you can also use a `let!` to that. There's no fundamental difference between this and the other `let!`s you defined. They are executed sequentially, like in any other imperative programming language, a line gets executed after the previous one. Therefore, it is guaranteed that `let!(:payment2)` gets executed after `let!(:order1)` if the order creation comes first.

What's more, this is a pattern you'll see often. Because `let(:payment)` references `order1` anyway, only turning `:payment` into a `let!` is enough to ensure things are executed in the right order. That's because the payment will be executed right away and the order will be called lazily just before that, because payment depends on it.

Reply:
Yes, that makes sense. However, if both payments for `order1` are defined with `let!` and evaluated sequentially, they will be evaluated outside the `it` blocks, which will make the first assertion fail:
line 119: `expect(page).to have_content "$88.00"`

As I understand this, it fails because it expects to find the `order1` in a state in which `payment2` was not yet introduced. Only the second assertion on this customer/order will pass:
line 146: `expect(page).to have_content "$63.00"`

Using `let` to define `payment2` and calling it when necessary was the only solution I could come up with to change `order1` and the balance of `customer1` from $88.00-> $63.00 inside the appropriate `it` block.

How to overcome this? Should we discuss this rather on Slack?

Reply:
I see. Yes, let's do a pair programming session. That'll be more efficient.

Reply:
Adding some notes here from our chat, for documentation purposes: I've added another commit, in which this test case is separated into two different `context`s. This allows to define `payment2` using `let!`, strictly inside the context and adding the `it`-block, which contains the necessary assertion. Hope this covers it appropriately. Thank you for your help!